### PR TITLE
Add ruff rules SIM to simplify Python code

### PR DESCRIPTION
**Pull request by cclauss**
https://docs.astral.sh/ruff/rules/#flake8-simplify-sim

% `ruff check --select=SIM --statistics`
```
26	SIM118	[*] in-dict-keys
 5	SIM105	[ ] suppressible-exception
 5	SIM108	[*] if-else-block-instead-of-if-exp
 2	SIM114	[*] if-with-same-arms
 2	SIM201	[*] negate-equal-op
 1	SIM101	[*] duplicate-isinstance-call
 1	SIM103	[*] needless-bool
```
% `ruff rule SIM118`
# in-dict-keys (SIM118)

Derived from the **flake8-simplify** linter.

Fix is always available.

## What it does
Checks for key-existence checks against `dict.keys()` calls.

## Why is this bad?
When checking for the existence of a key in a given dictionary, using
`key in dict` is more readable and efficient than `key in dict.keys()`,
while having the same semantics.

## Example
```python
key in foo.keys()
```

Use instead:
```python
key in foo
```

## Fix safety
Given `key in obj.keys()`, `obj` _could_ be a dictionary, or it could be
another type that defines a `.keys()` method. In the latter case, removing
the `.keys()` attribute could lead to a runtime error. The fix is marked
as safe when the type of `obj` is known to be a dictionary; otherwise, it
is marked as unsafe.

## References
- [Python documentation: Mapping Types](https://docs.python.org/3/library/stdtypes.html#mapping-types-dict)

[preview]: https://docs.astral.sh/ruff/preview/

### CODE REVIEW

File: icepyx/core/granules.py
Comment (mfisher87) [inline icepyx/core/granules.py:77]: Not a strong opinion, but I feel this is less readable despite having fewer lines.

File: icepyx/core/query.py
Comment (mfisher87) [inline icepyx/core/query.py:744]: Huge value from this type of simplification; it would be easy to introduce a bug by changing only one branch of the existing conditional and forgetting the other.

File: icepyx/quest/dataset_scripts/argo.py
Comment (mfisher87) [inline icepyx/quest/dataset_scripts/argo.py:297]: :100:

**Review by mfisher87 — CHANGES_REQUESTED** [review]
Love it! Except, and this is just my personal opinion, but I think the `contextlib.suppress` change is a (thin) obfuscation of the underlying `try`/`except` behavior and reduces readability. I can probably be convinced otherwise, but I've read a lot of Python and had to go to the docs to figure out what it did! [It doesn't seem uncommon to single this rule out for disabling](https://github.com/jupyter/jupyter_core/blob/fa513c1550bbd1ebcc14a4a79eb8c5d95e3e23c9/pyproject.toml#L183).

File: icepyx/core/granules.py
Comment (cclauss) [reply icepyx/core/granules.py:77]: It just takes getting used to it is easier to read once you do.

Comment (mfisher87) [reply icepyx/core/granules.py:77]: I think I'd need a bit more convincing on this :) Apologies, since I originally posted, my opinion has gotten a bit stronger. In my view, "taking some getting used to" is a really big cost! Code is read at least one order of magnitude more than it's written, so I care deeply about optimizing for folks who haven't gotten used to it yet.

**Comment by cclauss** [comment]
The suppressible-exception have been suppressed.

**Review by mfisher87 — APPROVED** [review]
LGTM! I'll leave some time for discussion before merging.

File: icepyx/core/APIformatting.py
Comment (JessicaS11) [inline icepyx/core/APIformatting.py:76]: I can see the argument for using this simplification (thanks for including the notes about it with the PR). However, I am concerned that this has potential to be confusing to folks newer to Python and ultimately make the code less clear/readable to them. It assumes an awareness that Python will automagically include `.keys()` if this is a dict, which is news to me as of this PR. What's the incentive for using this simplification over explicitly including `.keys()`?

Comment (mfisher87) [reply icepyx/core/APIformatting.py:76]: I think readability is paramount, and the performance concern is negligible, but want to mention it anyway. Calling `.keys()` is causing some extra overhead to create a [view](https://docs.python.org/3/library/stdtypes.html#dictionary-view-objects) of the dictionary's keys before checking.

While I do agree that calling `.keys()` here is more explicit, and [explicit is better than implicit](https://peps.python.org/pep-0020/), "readability counts". Sometimes the most readable thing is the way everyone else is doing it, and I think this is one of those cases. The official Python docs endorse this approach:

>To check whether a single key is in the dictionary, use the [in](https://docs.python.org/3/reference/expressions.html#in) keyword.

https://docs.python.org/3/tutorial/datastructures.html

Comment (cclauss) [reply icepyx/core/APIformatting.py:76]: Membership test operations (`item in container`) works for dicts, lists, sets, tuples, and all other classes that implement `.__contains__()`
* https://docs.python.org/3/reference/expressions.html#in
* [`https://docs.python.org/3/reference/datamodel.html#object.__contains__`](https://docs.python.org/3/reference/datamodel.html#object.__contains__)

Comment (mfisher87) [reply icepyx/core/APIformatting.py:76]: IMO this is a great use case for ChatGPT ("comprehension", in the "list-comprehension" sense, not the cognition sense, of documentation or community coding convention) so I threw this question at it, and it did kind of lie by misquoting the documentation, but it came pretty close to what I consider a good answer.

https://chatgpt.com/share/de22c27b-3370-4051-9d58-b5eeb37b9e09

**Comment by mfisher87** [comment]
>What I try to avoid is making complex manual changes in a pull request that is supposed to be based on automatic transforms. These make the PR too difficult to review and can introduce bugs that are not based on applying the rules of the tool. I know that the tool and its rules are imperfect but in general their transformations are useful.

:100: Agreed! But - when a rule does a thing that the project team doesn't universally agree on, IMO, that PR is the place to hash it out, and that's what I feel was happening in #571. I'm OK with addressing the concerns in #571 by disabling a rule and leaving a comment explaining why we disabled that rule (e.g. "automatic f-string conversion on this codebase makes some string interpolations harder to read") and opening an issue to track the desired steps. I'm also OK with mixing concerns a little bit (and what I mean by "a little bit" is based on vibes) - a PR that re-enabled that rule and also manually adjusted some strings would be fine, especially if those two things were in two separate commits I could review separately.

I think this PR is ready to merge if @JessicaS11 is now comfortable with our conversation above. If not, I think we should disable the rule which converts `... in foo.keys()` to `... in foo`, merge, and come back to it later.

> [!IMPORTANT]
>
> Going forward, let's please keep discussions about a PR associated with said PR!

File: icepyx/core/APIformatting.py
Comment (JessicaS11) [reply icepyx/core/APIformatting.py:76]: Thanks @cclauss and @mfisher87 for helping me understand this convention!

File: icepyx/core/variables.py
Comment (JessicaS11) [inline icepyx/core/variables.py:111]: Is this another case where there are "hidden" assumptions? Why doesn't this simplify to `return self._path if self._path else None`? To me, `return self._path or None` is like saying "yes" to "do you want an apple or an orange?" It's not obvious which one I should give you (and I can't give you both).

**Comment by JessicaS11** [comment]
>I think this PR is ready to merge if @JessicaS11 is now comfortable with our conversation above. If not, I think we should disable the rule which converts ... in foo.keys() to ... in foo, merge, and come back to it later.

Agreed. As a newbie to ruff - how do these "extras" fit into the Python formatting ecosystem? Some of these ruff rules are improving PEP compliance (yay!), some are aiming to make code a bit shorter (overall good), and some seem to be enforcing personal preferences that make the code shorter but I would argue less intuitive to read if you don't know the underlying assumptions. As a library that tries to be super-beginner friendly, I am trying to discern where it's acceptable to have a slightly "longer and less optimized" code base that is readable to a broader subset of scientific Python contributors. If I have to guess at what's happening under the hood (e.g. in the `return self._path or None` case), I can only imagine it will be the case for other developers who don't have "software" somewhere in their job title. This takes the ruff rule from "helpful" to "harmful" in my view.

File: icepyx/core/variables.py
Comment (cclauss) [reply icepyx/core/variables.py:111]: There are no assumptions as Truth Value Testing with `or` and `and` is defined quite clearly at the beginning of the Python Standard Library docs before `int`, `list`, `dict`, etc. are defined.

File: icepyx/core/APIformatting.py
Comment (mfisher87) [reply icepyx/core/APIformatting.py:76]: Happy to!

File: icepyx/core/variables.py
Comment (mfisher87) [reply icepyx/core/variables.py:111]: We can simplify this statement in two passes to make it a bit clearer:

```python
if self._path:
    path = self._path
else:
    path = None
return path
```

->

```python
if self._path:
    return self._path
else:
    return None
```

->

```python
return self._path or None
```

The `or` here is doing the lifting of the `else` condition -- if `self._path` is "Falsy", we move along to the next thing in the `or` operation.

Comment (cclauss) [reply icepyx/core/variables.py:111]:
```suggestion
        return self._path if self._path else None
```

**Comment by cclauss** [comment]
`return self._path if self._path else None` wins.

**Comment by mfisher87** [comment]
>how do these "extras" fit into the Python formatting ecosystem? Some of these ruff rules are improving PEP compliance (yay!), [...]

They largely come from the flake8 linting ecosystem. Flake8 has a huge number of plugins, and after a while I stopped thinking of it only in terms of PEP8 despite the name :)

> [...] some are aiming to make code a bit shorter (overall good), [...]

I try to be cautious with this, because sometimes shorter is better (more concise) and other times shorter introduces special concepts (like SIM105).

> [...] some seem to be enforcing personal preferences that make the code shorter but I would argue less intuitive to read if you don't know the underlying assumptions

Absolutely. The SIM105 rule is, to me, definitely on this side of the equation. In my experience, there are very few rules like this in the default set, but there are some.

**There's also a really important 4th category**: Things that aren't PEP8, aren't shorter, but are common pitfalls for Python developers. One of the most important rulesets, IMO, is the [flake8-bugbear](https://docs.astral.sh/ruff/rules/#flake8-bugbear-b) set. It detects, for example, attempting to clear environment vars by assigning to `os.environ` (B003). Or using a mutable object (B006) or calling a function (B008) to set a default argument. Things that look like will do what you want, but really are a bug or misleading. Then there's [flake8-unused-arguments](https://docs.astral.sh/ruff/rules/#flake8-unused-arguments-arg) which detects when you write a function with args but you don't use some of them in the function body, falsely advertising to users of the function that all of the args passed in are significant.

There are so many ways to write code that looks right, but isn't.

Type checking, to me, is another vital tool for detecting situations like this :point_up: and I'd like to get in to that space for v2. This is very valuable both for detecting bugs, but also for showing autocompletions in developers IDEs.

> If I have to guess at what's happening under the hood (e.g. in the `return self._path or None` case), I can only imagine it will be the case for other developers who don't have "software" somewhere in their job title. This takes the ruff rule from "helpful" to "harmful" in my view.

The functionality of `or`, I personally feel, is really important for beginner developers to understand; ["short-circuiting"](https://en.wikipedia.org/wiki/Short-circuit_evaluation) is very handy behavior. This is a "Pythonic" pattern that readers will definitely come across again, Ruff or no Ruff. That said, `return self._path if self._path else None` is trivially "less-Pythonic" in my view than `return self._path or None`.

**Comment by JessicaS11** [comment]
@all-contributors
please add @cclauss for maintenance
please add @mfisher87 for bug, code, doc, maintenance, test, infra
please add @jrenrut for bug, code

**Comment by JessicaS11** [comment]
Thanks for this great response! I'm learning so much.

> The functionality of or, I personally feel, is really important for beginner developers to understand; ["short-circuiting"](https://en.wikipedia.org/wiki/Short-circuit_evaluation) is very handy behavior.

I just taught a bunch of folks about this concept, so it was exciting to learn with them. I agree it's a good feature to know about (and @RomiP provided some helpful insight on its ability to make code more efficient by not running unnecessary bits).

**Review by JessicaS11 — APPROVED** [review]
Thanks for this PR and all the great discussion!

**Comment by mfisher87** [comment]
@all-contributors
please add @cclauss for review

**Comment by mfisher87** [comment]
@all-contributors
please add @jrenrut for bug, code

**Pull request merged**
